src/test/java/aeneas/views/PieceManipulationTest.java

```diff
@@ -64,46 +73,79 @@ private PieceView getFirstPiece() {
     press(KeyCode.SHIFT).clickOn(piece).release(KeyCode.SHIFT);
     oldPiece.rotate(Piece.Dir.COUNTERCLOCKWISE);
     assertEquals(oldPiece.toString(), piece.pieceModel.toString());
   }
 
   @Test
   public void testFlipPiece() {
-    navigateToFirstLevel();
+    navigateToLevel(1);
     PieceView piece = getFirstPiece();
     Piece oldPiece = piece.pieceModel.clone();
     // Now, test flips.
     press(KeyCode.CONTROL).clickOn(piece).release(KeyCode.CONTROL);
     oldPiece.flip(Piece.Axis.VERTICAL);
     assertEquals(oldPiece.toString(), piece.pieceModel.toString());
 
-    System.out.println("Horizontal: ");
     press(KeyCode.CONTROL).press(KeyCode.SHIFT).clickOn(piece).release(KeyCode.CONTROL).release(KeyCode.SHIFT);
     oldPiece.flip(Piece.Axis.HORIZONTAL);
     assertEquals(oldPiece.toString(), piece.pieceModel.toString());
   }
 
-  @Test
-  public void testDragPiece() {
-    navigateToFirstLevel();
-    VBox bullpenBox = (VBox) lookup("#bullpenBox").query();
-    Pane piecePane = (Pane) bullpenBox.getChildren().get(0);
-    PieceView piece = (PieceView) piecePane.getChildren().get(0);
+  BoardView boardView;
+  VBox bullpenBox;
+
+  public void testDragPieceSetup(int level_num) {
+    navigateToLevel(level_num);
+    bullpenBox = (VBox) lookup("#bullpenBox").query();
     VBox boardBox = (VBox) lookup("#centerBox").query();
     ObservableList<Node> children = boardBox.getChildren();
-    BoardView boardView = (BoardView) children.get(children.size()-1);
+    boardView = (BoardView) children.get(children.size()-1);
+  }
+
+  @Test
+  public void testDragPuzzle() {
+    testDragPieceSetup(1);
+
+    Pane piecePane = (Pane) bullpenBox.getChildren().get(0);
+    // First, drag piece to board.
+    drag(piecePane.getChildren().get(0)).dropTo(boardView.grid[0][0]);
+    assertEquals(1, boardView.board.getPieces().size());
+
+    // Now, drag piece back to Bullpen.
+    drag(boardView.grid[0][0]).dropTo(bullpenBox);
+    assertEquals(0, boardView.board.getPieces().size());
 
+    piecePane = (Pane) bullpenBox.getChildren().get(0);
+    // Now, drag to invalid spot on board.
+    drag(piecePane.getChildren().get(0)).dropTo(boardView.grid[11][11]);
+    assertEquals(0, boardView.board.getPieces().size());
+  }
+
+  @Test
+  public void testDragLightning() throws InterruptedException {
+    testDragPieceSetup(2);
+
+    Pane piecePane = (Pane) bullpenBox.getChildren().get(0);
+    // First, drag piece to board.
+    drag(piecePane.getChildren().get(0)).dropTo(boardView.grid[0][0]);
+    assertEquals(0, boardView.board.getPieces().size());
+  }
+
+  @Test
+  public void testDragRelease() {
+    testDragPieceSetup(3);
+
+    Pane piecePane = (Pane) bullpenBox.getChildren().get(0);
     // First, drag piece to board.
-    drag(piece).dropTo(boardView.grid[0][0]);
+    drag(piecePane.getChildren().get(0)).dropTo(boardView.grid[0][0]);
     assertEquals(1, boardView.board.getPieces().size());
 
     // Now, drag piece back to Bullpen.
```

Comment:
I just realized that in release you aren't supposed to be able to remove pieces from the board. Maybe we do need BoardLogic.